document how to generate vendor-agnostic SQL with sqlglot (#371)
ploomber/jupysql: doc/community/developer-guide.md

```{code-cell} ipython3
assert str(excinfo.value) == "division by zero"
```

## General SQL Clause for Multiple Database Dialects

### Context

As our codebase is expanding, we have noticed that we need to write SQL queries for different database dialects such as MySQL, PostgreSQL, SQLite, and more. Writing and maintaining separate queries for each database can be time-consuming and error-prone.

> To address this issue, we have created a general SQL clause that can be used across multiple database dialects. This clause will allow us to write a single SQL query that can be translated to different database dialects, then use it for calculating the metadata (e.g. metadata used by boxplot)

Reviewer: ```suggestion
To address this issue, we have created a general SQL clause that can be used across multiple database dialects. This clause will allow us to write a single SQL query that can be translated to different database dialects.
```

> To address this issue, we have created a general SQL clause that can be used across multiple database dialects. This clause will allow us to write a single SQL query that can be translated to different database dialects, then use it for calculating the metadata (e.g. metadata used by boxplot)

Reviewer: ```suggestion
To address this issue, we can use `sqlglot` to create a construct that can be compiled across multiple SQL dialects. This clause will allow us to write a single SQL query that can be translated to different database dialects, then use it for calculating the metadata (e.g. metadata used by boxplot)
```

> In this document, we will provide an overview of the general SQL clause, explain how it works, and provide examples of how it can be used in our codebase. We will also include instructions on how to add support for additional database dialects.

Reviewer: ```suggestion
In this document, we'll explain how to build generic SQL constructs and provide examples of how it can be used in our codebase. We will also include instructions on how to add support for additional database dialects.
```

### Example

We can use [SQLGlot](https://sqlglot.com/sqlglot.html) to build the general sql expressions

> 

Reviewer: use

~~~
```{code-cell} ipython3
````
~~~

so this becomes a runnable cell and executes while building the docs:

https://ploomber-contributing.readthedocs.io/en/latest/documentation/notebooks.html

```python
from sqlglot import select, condition

where = condition("x=1").and_("y=1")

> general_sql = select("*").from_("y").where(where).sql()

Reviewer: profile  a full example where we take this `general_sql` and convert it into several dialects. perhaps create a session with a sqlite and a duckdb connection to show how the transpile query call converts to the current active dialect

```
```python
'SELECT x FROM y, z'
```

Then transpile to the sql which is supported by current connected dialect.

Our `sql.connection.Connection._transpile_query` will automatically detect the dialect and transpile the SQL clause

```python

> query = sql.connection.Connection._transpile_query(general_sql)

Reviewer: does the transpile_query function behave differently if it receives a string vs a sqlglot object? I'd assume that passing a sqlglot object is more reliable since it should be easier to export to certain dialect as we're removing the initial parsing stage

Author: I did the experiment and figure out our current `transpile_query()` can only help us to solve the "`" (backtick) transformation issue

In order to see the different transpiled sql results under different dialects, we need to provide the original SQL Clause with original dialect, Something like this:

```
from sql import connection
from sqlalchemy import create_engine

# Prepare connection
conn = connection.Connection(engine=create_engine(url="sqlite://"))

# Prepare SQL clause based on duckdb syntax
input_sql = sqlglot.parse_one("SELECT TO_TIMESTAMP(1618088028295)", read="duckdb").sql()
print(f"{conn._get_curr_sqlglot_dialect()} transpiled result:")
conn._transpile_query(input_sql)

# sqlite transpiled result:
# 'SELECT UNIX_TO_TIME(1618088028295)'
```

We run integration tests against cloud databases like Snowflake, which requires using pre-registered accounts to evaluate their behavior. To initiate these tests, please create a branch in our [ploomber/jupyter repository](https://github.com/ploomber/jupysql).

Please note that if you submit a pull request from a forked repository, the integration testing phase will be skipped because the pre-registered accounts won't be accessible.
## General SQL Clause for Multiple Database Dialects

### Context

As our codebase is expanding, we have noticed that we need to write SQL queries for different database dialects such as MySQL, PostgreSQL, SQLite, and more. Writing and maintaining separate queries for each database can be time-consuming and error-prone.

To address this issue, we can use `sqlglot` to create a construct that can be compiled across multiple SQL dialects. This clause will allow us to write a single SQL query that can be translated to different database dialects, then use it for calculating the metadata (e.g. metadata used by boxplot)

> In this document, we'll explain how to build generic SQL constructs and provide examples of how it can be used in our codebase. We will also include instructions on how to add support for additional database dialects.

Reviewer 2: Suggestion : 
in this **section** instead in this **document**?

### Approach 1 - Provide the general SQL Clause

We can use [SQLGlot](https://sqlglot.com/sqlglot.html) to build the general sql expressions.

Then transpile to the sql which is supported by current connected dialect.

Our `sql.connection.Connection._transpile_query` will automatically detect the dialect and transpile the SQL clause.

+++

#### Example

```{code-cell} ipython3
from sqlglot import select, condition
from sql import connection
from sqlalchemy import create_engine

# Prepare connection
conn = connection.Connection(engine=create_engine(url="sqlite://"))

where = condition("x=1").and_("y=1")
general_sql = select("*").from_("y").where(where).sql()

print("General SQL Clause: ")
print(f"{general_sql}\n")

print("Transpiled result: ")

> conn._transpile_query(general_sql)

Reviewer 2: ![image](https://user-images.githubusercontent.com/103026689/232440257-17958a80-12aa-41b4-8825-23a9c6c7e477.png)

I think this might be hard to read this output cell.

What do you think if we'll split it into 3 code cells, where the 1st cell initializes the connection, 2nd cell constructs the SQL clause, and 3rd cell transpiles the query. We can drop the prints and use headers above each cell.

Author: This is already implemented here?
https://github.com/ploomber/jupysql/pull/371#issuecomment-1511466470

```

### Approach 2 - Provide SQL Clause based on specfic database 

Sometimes the SQL Clause might be complex, we can also write the SQL Clause based on one specfic database and transpile it.

For example, the `TO_TIMESTAMP` keyword is only defined in duckdb, but we want to also apply this SQL clause to other database.

We may provide `sqlglot.parse_one({source_sql_clause}, read={source_database_dialect}).sql()` as input sql to `_transpile_query()`

+++

#### When current connection is via duckdb

```{code-cell} ipython3
from sql import connection
from sqlalchemy import create_engine
import sqlglot

# Prepare connection
conn = connection.Connection(engine=create_engine(url="duckdb://"))

# Prepare SQL clause based on duckdb syntax
input_sql = sqlglot.parse_one("SELECT TO_TIMESTAMP(1618088028295)", read="duckdb").sql()

print(f"{conn._get_curr_sqlglot_dialect()} transpiled result:")

> conn._transpile_query(input_sql)

Reviewer 2: ![image](https://user-images.githubusercontent.com/103026689/232441443-e28b8b53-f684-4bd7-9a82-8a5b72dc0e91.png)

same as before. I think we can remove the print statement OR include `conn._transpile_query(input_sql)` in the print OR out the last command `conn._transpile_query(input_sql)` in a new cell with title `duckdb transpiled result`

```

#### When current connection is via sqlite

```{code-cell} ipython3
from sql import connection
from sqlalchemy import create_engine

# Prepare connection
conn = connection.Connection(engine=create_engine(url="sqlite://"))

# Prepare SQL clause based on duckdb syntax
input_sql = sqlglot.parse_one("SELECT TO_TIMESTAMP(1618088028295)", read="duckdb").sql()

print(f"{conn._get_curr_sqlglot_dialect()} transpiled result:")

> conn._transpile_query(input_sql)

Reviewer 2: Same as above

```

As you can see, output results are different

From duckdb dialect: `'SELECT TO_TIMESTAMP(1618088028295)'`

From sqlite dialect: `'SELECT UNIX_TO_TIME(1618088028295)'`

> ```{code-cell} ipython3

Reviewer 2: Can we remove this empty code cell?
